web/src/recent_view_ui.ts

```diff
@@ -794,6 +794,27 @@ export function update_topics_of_deleted_message_ids(message_ids: number[]): voi
         msgs_to_process.push(...msgs);
     }
 
+    const unique_dm_keys_to_repopulate = new Set<string>();
+    for (const msg_id of message_ids) {
+        const msg = message_store.get(msg_id);
+        if (msg === undefined) {
+            continue;
+        }
+
+        if (msg.type === "private") {
+            const key = recent_view_util.get_key_from_message(msg);
+            // Important to assert since we use the key in get_messages_in_dm.
+            assert(key === msg.to_user_ids);
+            unique_dm_keys_to_repopulate.add(key);
+        }
+    }
+
+    for (const key of unique_dm_keys_to_repopulate) {
+        recent_view_data.conversations.delete(key);
+        const msgs_in_dm = message_util.get_messages_in_dm(key);
+        msgs_to_process.push(...msgs_in_dm);
+    }
```

Comment:
This is still quadratic if lots of messages were deleted at once, say via messages retention policy, though much less likely to be relevant.

Fixed by changing the APIs a bit.